src/oauthcli/storage.py

```diff
@@ -0,0 +1,94 @@
+"""Tokens storage"""
+
+import json
+import logging
+import os
+from typing import Optional
+
+import keyring
+import platformdirs
+
+log = logging.getLogger(__name__)
+
+
+class BaseStorage:
+
+    @staticmethod
+    def _token_key(provider_id: str, client_id: str):
+        return f'{provider_id}/{client_id}'
+
+    def load(self):
+        raise NotImplementedError()
+
+    def save(self, tokens: dict):
+        raise NotImplementedError()
+
+    def get_token(self, provider_id: str, client_id: str):
+        return self.load().get(self._token_key(provider_id, client_id))
+
+    def set_token(self, provider_id: str, client_id: str, token: Optional[dict]):
+        token_key = self._token_key(provider_id, client_id)
+        tokens = self.load()
+
+        if not token:
+            if token_key in tokens:
+                del tokens[token_key]
+        else:
+            tokens[token_key] = token
+        self.save(tokens)
+
+
+class ConfigFileStorage(BaseStorage):
+
+    def __init__(self, path: Optional[str] = None, config_name=None):
+        self.path = (
+            path if path
+            else os.path.join(
+                platformdirs.user_config_dir(config_name if config_name else 'PythonCliAuth', ensure_exists=True),
+                'tokens.json'
+            )
+        )
+
+    def load(self):
+        log.debug("Load tokens from %s", self.path)
+        if not os.path.exists(self.path):
+            log.debug("Tokens file %s doest not exists", self.path)
+            return {}
+        with open(self.path, 'r') as f:
+            return json.load(f)
+
+    def save(self, tokens: dict):
+        log.debug("Save tokens to %s", self.path)
+        try:
+            with open(self.path, 'w') as f:
+                json.dump(tokens, f)
+        except IOError:
+            log.exception('Error saving tokens to %s', self.path)
+
+
+class KeyringStorage(BaseStorage):
+
+    def __init__(self, service_name: Optional[str] = None):
```

Comment:
It looks weird that you specify `config_name` for the file class and `service_name` for this one. Maybe unify those, and use a constant for the default? Or this one should not have title case?

Reply:
Parameters could really differ from a storage to another, so it's could be a bad idea to try to unify it to me. There, I only add a "name" as parameter, but for instance, for the keyring backend, we could also have a `backend` parameter to force using a specify keyring backend. Since we have to instantiate them independently from the `AuthFlow`, it doesn't seem problematic to me that the parameters differ from one storage to another. But say me if you have an idea to how to handle different parameters in a more common way.